bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -25,219 +24,396 @@ import jsky.app.ot.gemini.tpe.TpePWFSFeature
-                      LOG.warning(s"$bagsIdMsg failed: ${e.msg}")
-                      enqueue(obs, 5000L)
-
-                    // If we timed out, we don't want to delay.
-                    case Failure(ex: TimeoutException) =>
-                      LOG.warning(s"$bagsIdMsg failed: ${ex.getMessage}")
-                      enqueue(obs, 0L)
-
-                    // For all other exceptions, print the full stack trace.
-                    case Failure(ex) =>
-                      LOG.log(Level.WARNING, s"$bagsIdMsg} failed.", ex)
-                      enqueue(obs, 5000L)
-                  }
-                }
-              } else if (!eligibleForBags) {
-                LOG.info(s"${obs.getObservationID} not eligible for BAGS. Clearing auto group.")
-                applySwingResults(None)
-              }
-            }
-          }
-        }
-      }
+  def apply(p: ISPProgram): ProgKey =
+    ProgKey(p.getProgramKey)
+
+  implicit val OrderProgKey: Order[ProgKey] =
+    Order.order((k0,k1) => Ordering.fromInt(k0.k.compareTo(k1.k)))
+}
+
+// Wraps SPNodeKey in a new type to help avoid confusion with program SPNodeKey.
+final case class ObsKey(k: SPNodeKey)
+object ObsKey {
+  def apply(o: ISPObservation): ObsKey =
+    ObsKey(o.getNodeKey)
+
+  implicit val OrderObsKey: Order[ObsKey] =
+    Order.order((k0,k1) => Ordering.fromInt(k0.k.compareTo(k1.k)))
+}
+
+
+object BagsManager {
+  private val Log = Logger.getLogger(getClass.getName)
+
+  // Worker pool running the AGS search and the timer.
+  val NumWorkers = math.max(1, Runtime.getRuntime.availableProcessors - 1)
+  private val worker = new ScheduledThreadPoolExecutor(NumWorkers, new ThreadFactory() {
+    override def newThread(r: Runnable): Thread = {
+      val t = new Thread(r, "BagsManager - Worker")
+      t.setPriority(Thread.NORM_PRIORITY - 1)
+      t.setDaemon(true)
+      t
+    }
+  })
+  private implicit val executionContext = ExecutionContext.fromExecutor(worker)
+
+  // This is our mutable state.  It is only read/written by the Swing thread.
+  private var stateMap  = ==>>.empty[ProgKey, ObsKey ==>> BagsState]
+
+  // Handle a state machine transition.  Note we switch to the Swing thread
+  // here so that the calling thread continues immediately.  All updates are
+  // done from the Swing thread.
+  private def transition(obs: ISPObservation, update: BagsState => StateTransition): Unit = Swing.onEDT {
+    val progKey = ProgKey(obs)
+    val obsKey  = ObsKey(obs)
+
+    val stateTuple = for {
+      obsMap <- stateMap.lookup(progKey)
+      state  <- obsMap.lookup(obsKey)
+    } yield (obsMap, state)
+
+    stateTuple.foreach { case (obsMap, state) =>
+      val (newState, sideEffect) = update(state)
+      val newStatemap            = stateMap.insert(progKey, obsMap.insert(obsKey, newState))
+
+      val action = for {
+        _ <- IO { stateMap = newStatemap }
+        _ <- sideEffect
+      } yield ()
+
+      action.unsafePerformIO()
```

Comment:
There is an action that will trigger an AGS search, but it will be done in a separate thread.  See `triggerAgsAction`.  There `ags.select` returns a `Future`.  If the future gets a successful result, it will make a `succeed` event which from `RunningState` moves to `IdleState`, etc.  If it fails, it will make a `fail` event, etc.